sklearn/preprocessing/_csr_polynomial_expansion.pyx

```diff
@@ -26,8 +39,13 @@ cdef inline INDEX_T _deg2_column(INDEX_T d, INDEX_T i, INDEX_T j,
         return d * i - (i**2 + i) / 2 + j
 
 
-cdef inline INDEX_T _deg3_column(INDEX_T d, INDEX_T i, INDEX_T j, INDEX_T k,
-                                 INDEX_T interaction_only) nogil:
+cdef inline cnp.int64_t _deg3_column(
+    cnp.int64_t d,
+    cnp.int64_t i,
+    cnp.int64_t j,
+    cnp.int64_t k,
+    FLAG_t interaction_only
+    ) nogil:
```

Comment:
I think the greatest pain point of the current implementation is that in order to compute the output index we must compute squares or cubes of indices (e.g. `i,j,k`) and while the output index may fit within `int{32,64}`, the intermediate square/cube calculation may overflow, throwing the entire thing off. This effectively limits the valid range of input indices.

I'm open to any suggestions for how to circumvent this issue.

Reply:
I think we can try using intermediary `int128`-typed variables for squared or cubes of indices.

NumPy does not defines `int128_t` but does define [`npy_int128`](https://github.com/numpy/numpy/blob/bb95cf015165bcf85f99ba0d28a7a6e727ca955b/numpy/__init__.pxd#L276) which should be available via `cnp.npy_int128` in Cython.

Reply:
Alright I ended up predicting when an intermediate overflow would occur, and then deferring to python in those cases (to make use of arbitrary-precision `PyLong`s). This _only_ occurs for the cases where the indices are too large to safely compute the output index, and so won't kick into effect unless the input data has billions of features in the best case, and 500k in the worst case. Even when it kicks in, the performance drop only affects exactly the elements with too-large indices, which I think is reasonable to assume would be the minority of the data.

<details>
<summary> Quick benchmarks: </summary>

```python
from sklearn.preprocessing import PolynomialFeatures
from scipy import sparse

X = sparse.random(100_000, 1000, random_state=0)
pf = PolynomialFeatures(interaction_only=False, include_bias=True, degree=2)
%timeit -n9 pf.fit_transform(X)
```
```
Cython only: 774 ms ± 7.6 ms per loop (mean ± std. dev. of 7 runs, 9 loops each)
PR:          784 ms ± 4.3 ms per loop (mean ± std. dev. of 7 runs, 9 loops each)
```
</details>

Reply:
Unfortunately the PR causes some serious GIL contention as demonstrated in the follow new version of the benchmark:

```python
from sklearn.preprocessing import PolynomialFeatures
from scipy import sparse
from concurrent.futures import ThreadPoolExecutor

tpe = ThreadPoolExecutor(max_workers=8)

X = sparse.random(100_000, 1000, random_state=0)
pf = PolynomialFeatures(interaction_only=False, include_bias=True, degree=2)
%timeit -n 3 list(tpe.map(pf.fit_transform, [X] * 2 * 8))
``` 

```
main: 793 ms ± 11 ms per loop (mean ± std. dev. of 7 runs, 3 loops each)
PR:   2.28 s ± 21.9 ms per loop (mean ± std. dev. of 7 runs, 3 loops each)
```

on a laptop with 8 physical cores.

I haven't checked the math but maybe there is a way to be more conservative w.r.t. using the Python arbitrary precision code?

Reply:
Or maybe we could go back to using `float64` arithmetic operations before a final integer cast to int64 as was done previously? Not sure if the rounding errors/lack of exact representability of all integers in the `float64` domain can be observed for this case.

Reply:
> As NumPy does not defines `int128_t` and does define [`npy_int128` to be identical to `int64_t`](https://github.com/numpy/numpy/blob/bb95cf015165bcf85f99ba0d28a7a6e727ca955b/numpy/__init__.pxd#L276), I think we can to compute squared or cubes of indices try using `np.uint64_t`-typed variables (which are `unsigned long long int`-typed variables)
> 
> Alternatively, we could try to rework the formula not to have huge values.

I'm sure the formulas could be rewritten to be more computationally-friendly. I was going for simplicity.

Reply:
> Unfortunately the PR causes some serious GIL contention as demonstrated in the follow new version of the benchmark:

I'm not sure that the original problem was indeed a GIL contention, rather just a general slowdown due to the introduction of `_calc_expanded_dimensions` which was an expensive and repeated calculation done entirely in Python. I've since moved it into the Cython layer (and renamed to `_calc_expanded_nnz`). Now the times of your script are as follow:

```
main: 642 ms ± 25.3 ms per loop (mean ± std. dev. of 7 runs, 3 loops each)
PR: 644 ms ± 10.2 ms per loop (mean ± std. dev. of 7 runs, 3 loops each)
```